File: packages/domain/src/timing.ts
import {
  MAX_DURATION_SECONDS,
  MAX_EPOCHS,
  MAX_EPOCH_SECONDS,
  MIN_EPOCH_SECONDS,
} from "./constants.ts";
import { fail } from "./errors.ts";
import { asI64 } from "./math.ts";

export interface ScheduleBounds {
  readonly minEpochSeconds: bigint;
  readonly maxEpochSeconds: bigint;
  readonly maxDurationSeconds: bigint;
  readonly maxEpochs: number;
}

export const DEFAULT_SCHEDULE_BOUNDS: ScheduleBounds = {
  minEpochSeconds: MIN_EPOCH_SECONDS,
  maxEpochSeconds: MAX_EPOCH_SECONDS,
  maxDurationSeconds: MAX_DURATION_SECONDS,
  maxEpochs: MAX_EPOCHS,
};

export interface Schedule {
  readonly startAt: bigint;
  readonly epochSeconds: bigint;
  readonly totalEpochs: number;
  readonly endAt: bigint;
}

/**
 * Build the fixed epoch schedule. Duration must be an exact multiple of the epoch length (v1).
 * Rule order is part of the contract: golden vectors assert which error wins.
 */
export function buildSchedule(
  startAt: bigint,
  durationSeconds: bigint,
  epochSeconds: bigint,
  bounds: ScheduleBounds = DEFAULT_SCHEDULE_BOUNDS,
): Schedule {
  if (epochSeconds < bounds.minEpochSeconds || epochSeconds > bounds.maxEpochSeconds) {
    fail("InvalidEpochLength", "epoch length outside protocol bounds");
  }
  if (durationSeconds <= 0n || durationSeconds > bounds.maxDurationSeconds || startAt <= 0n) {
    fail("InvalidTiming", "duration or start outside bounds");
  }
  if (durationSeconds % epochSeconds !== 0n) {
    fail("InvalidEpochLength", "duration is not a multiple of the epoch length");
  }
  const total = durationSeconds / epochSeconds;
  if (total > BigInt(bounds.maxEpochs)) fail("TooManyEpochs");
  const endAt = asI64(startAt + epochSeconds * total);
  return { startAt, epochSeconds, totalEpochs: Number(total), endAt };
}

export type EpochPosition =
  | { readonly kind: "NotStarted" }
  | { readonly kind: "Ended" }
  | { readonly kind: "Epoch"; readonly index: number };

/**
 * Which epoch a timestamp falls in. Epoch `i` covers `[start + i*E, start + (i+1)*E)`: the start
 * is inclusive and the end exclusive, so a timestamp exactly on a boundary belongs to the later epoch.
 */
export function epochPositionAt(
  schedule: Pick<Schedule, "startAt" | "epochSeconds" | "totalEpochs">,
  ts: bigint,
): EpochPosition {
  const endAt = schedule.startAt + schedule.epochSeconds * BigInt(schedule.totalEpochs);
  if (ts < schedule.startAt) return { kind: "NotStarted" };
  if (ts >= endAt) return { kind: "Ended" };
  return { kind: "Epoch", index: Number((ts - schedule.startAt) / schedule.epochSeconds) };
}

export interface EpochBounds {
  readonly epochStart: bigint;
  /** Exclusive end. An epoch can be finalized once the clock reaches this instant. */
  readonly epochEnd: bigint;
  /** Earliest instant `finalize_unavailable_epoch` may run (epoch end + recovery window). */
  readonly recoveryDeadline: bigint;
}

export function epochBounds(
  schedule: Pick<Schedule, "startAt" | "epochSeconds" | "totalEpochs">,
  index: number,
  unavailableRecoverySeconds: bigint,
): EpochBounds {
  if (!Number.isSafeInteger(index) || index < 0 || index >= schedule.totalEpochs)
    fail("EpochOutOfRange");
  const epochStart = schedule.startAt + BigInt(index) * schedule.epochSeconds;
  const epochEnd = epochStart + schedule.epochSeconds;
  return { epochStart, epochEnd, recoveryDeadline: asI64(epochEnd + unavailableRecoverySeconds) };
}

/**
 * The last instant a sponsor can still accept a bid: `startAt - lockBuffer - setupWindow`. Accepting no later
 * than this leaves the provider at least the protocol's setup window before the position set locks
 * (docs/adr/0009). Throws `ArithmeticOverflow` rather than leaving the i64 range.
 */
export function acceptanceCutoff(
  startAt: bigint,
  positionLockBufferSeconds: bigint,
  minSetupWindowSeconds: bigint,
): bigint {
  return asI64(startAt - positionLockBufferSeconds - minSetupWindowSeconds);
}

File: crates/mandate-core/src/timing.rs
//! Epoch schedule and timing. Epoch `i` covers `[start + i*E, start + (i+1)*E)`: start inclusive,
//! end exclusive, so a timestamp on a boundary belongs to the later epoch.
use crate::math::narrow_i64;
use crate::{
    MandateCoreError, Result, MAX_DURATION_SECONDS, MAX_EPOCHS, MAX_EPOCH_SECONDS,
    MIN_EPOCH_SECONDS,
};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct ScheduleBounds {
    pub min_epoch_seconds: i64,
    pub max_epoch_seconds: i64,
    pub max_duration_seconds: i64,
    pub max_epochs: u32,
}

pub const DEFAULT_SCHEDULE_BOUNDS: ScheduleBounds = ScheduleBounds {
    min_epoch_seconds: MIN_EPOCH_SECONDS,
    max_epoch_seconds: MAX_EPOCH_SECONDS,
    max_duration_seconds: MAX_DURATION_SECONDS,
    max_epochs: MAX_EPOCHS,
};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Schedule {
    pub start_at: i64,
    pub epoch_seconds: i64,
    pub total_epochs: u32,
    pub end_at: i64,
}

/// Build the fixed schedule. Rule order is part of the contract (golden vectors assert which error wins).
pub fn build_schedule(
    start_at: i64,
    duration_seconds: i64,
    epoch_seconds: i64,
    bounds: &ScheduleBounds,
) -> Result<Schedule> {
    if epoch_seconds < bounds.min_epoch_seconds
        || epoch_seconds > bounds.max_epoch_seconds
        || epoch_seconds <= 0
    {
        return Err(MandateCoreError::InvalidEpochLength);
    }
    if duration_seconds <= 0 || duration_seconds > bounds.max_duration_seconds || start_at <= 0 {
        return Err(MandateCoreError::InvalidTiming);
    }
    let remainder = duration_seconds
        .checked_rem(epoch_seconds)
        .ok_or(MandateCoreError::InvalidEpochLength)?;
    if remainder != 0 {
        return Err(MandateCoreError::InvalidEpochLength);
    }
    let total = duration_seconds
        .checked_div(epoch_seconds)
        .ok_or(MandateCoreError::InvalidEpochLength)?;
    if total > i64::from(bounds.max_epochs) {
        return Err(MandateCoreError::TooManyEpochs);
    }
    let end_at = epoch_seconds
        .checked_mul(total)
        .and_then(|span| start_at.checked_add(span))
        .ok_or(MandateCoreError::ArithmeticOverflow)?;
    let total_epochs = u32::try_from(total).map_err(|_| MandateCoreError::TooManyEpochs)?;
    Ok(Schedule {
        start_at,
        epoch_seconds,
        total_epochs,
        end_at,
    })
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum EpochPosition {
    NotStarted,
    Ended,
    Epoch(u32),
}

/// Which epoch a timestamp falls in.
pub fn epoch_position_at(
    start_at: i64,
    epoch_seconds: i64,
    total_epochs: u32,
    ts: i64,
) -> Result<EpochPosition> {
    let span = i128::from(epoch_seconds)
        .checked_mul(i128::from(total_epochs))
        .ok_or(MandateCoreError::ArithmeticOverflow)?;
    let end_at = i128::from(start_at)
        .checked_add(span)
        .ok_or(MandateCoreError::ArithmeticOverflow)?;
    if ts < start_at {
        return Ok(EpochPosition::NotStarted);
    }
    if i128::from(ts) >= end_at {
        return Ok(EpochPosition::Ended);
    }
    let elapsed = i128::from(ts)
        .checked_sub(i128::from(start_at))
        .ok_or(MandateCoreError::ArithmeticOverflow)?;
    let index = elapsed
        .checked_div(i128::from(epoch_seconds))
        .ok_or(MandateCoreError::InvalidEpochLength)?;
    let index = u32::try_from(index).map_err(|_| MandateCoreError::ArithmeticOverflow)?;
    Ok(EpochPosition::Epoch(index))
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct EpochBounds {
    pub epoch_start: i64,
    /// Exclusive end. An epoch can be finalized once the clock reaches this instant.
    pub epoch_end: i64,
    /// Earliest instant an unavailable finalization may run: epoch end plus the recovery window.
    pub recovery_deadline: i64,
}

pub fn epoch_bounds(
    start_at: i64,
    epoch_seconds: i64,
    total_epochs: u32,
    index: u32,
    unavailable_recovery_seconds: i64,
) -> Result<EpochBounds> {
    if index >= total_epochs {
        return Err(MandateCoreError::EpochOutOfRange);
    }
    let offset = i128::from(index)
        .checked_mul(i128::from(epoch_seconds))
        .ok_or(MandateCoreError::ArithmeticOverflow)?;
    let epoch_start = i128::from(start_at)
        .checked_add(offset)
        .ok_or(MandateCoreError::ArithmeticOverflow)?;
    let epoch_end = epoch_start
        .checked_add(i128::from(epoch_seconds))
        .ok_or(MandateCoreError::ArithmeticOverflow)?;
    let recovery_deadline = epoch_end
        .checked_add(i128::from(unavailable_recovery_seconds))
        .ok_or(MandateCoreError::ArithmeticOverflow)?;
    Ok(EpochBounds {
        epoch_start: narrow_i64(epoch_start)?,
        epoch_end: narrow_i64(epoch_end)?,
        recovery_deadline: narrow_i64(recovery_deadline)?,
    })
}
